### PR TITLE
Docs: Versioning long running functions (and determinism) explanation

### PR DESCRIPTION
**Pull request by djfarrelly**
Extracted and edited from [this guide draft](https://www.notion.so/inngest/Running-in-production-690950f872ba4d2ca7613c368d699ce7?pvs=4#5984c8b42c73461dbff1eccdb7a14683) to be a standalone explanation guide on how determinism works and share approaches for change management. 

My thoughts on extracting this from the drafted "Running in production" guide is that this is a generally applicable concept to understand, not just in production. This also compliments the "how inngest functions are executed" doc in #760. This doc might be better suited in guides, but my thought is that this is more of an explanation which feels more appropriate adjacent to the other "how inngest works" docs.

I removed some parts from the original that are not yet accurate and may be handled in future versions.

### CODE REVIEW

File: pages/docs/learn/versioning.mdx [hunk 1]
```diff
@@ -0,0 +1,39 @@
+export const description = "Learn how ."
+
+# Versioning long running functions
+
+It's inevitable that long-running functions change over time. Inngest allows developers to implement multiple strategies for changing long-running code over time. In order to understand how to best manage changes over time, it's important to understand how the SDK implements determinism and [executes steps](/docs/learn/how-functions-are-executed).
```
Comment (sylwiavargas) [inline pages/docs/learn/versioning.mdx:5]:
```suggestion
Long-running functions inevitably change over time. Inngest enables developers to implement multiple strategies for changing long-running code over time. To manage these changes effectively, it's crucial to understand how the SDK implements determinism and [executes steps](https://chatgpt.com/docs/learn/how-functions-are-executed).
```

- changes the first sentence to be clearer and more impactful
- solves redundant repetition in the last sentence

File: pages/docs/learn/versioning.mdx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Determinism in functions
+
+Determinism is the same in *every* Inngest language SDK. Except for language-specific idioms, all SDKs implement the same logic. In all SDKs, functions in Inngest are a series of steps. Each step runs reliably, will retry on failure, and is as close to exactly once as possible (outbound network failure when reporting completing steps notwithstanding).
```
Comment (sylwiavargas) [inline pages/docs/learn/versioning.mdx:9]:
```suggestion
Determinism is consistent in *every* Inngest language SDK. Except for language-specific idioms, all SDKs implement the same logic. In every SDK, functions in Inngest are a series of steps. Each step runs reliably, will retry on failure, and is as close to exactly-once execution as possible (excluding outbound network failure when reporting completing steps).
```

File: pages/docs/learn/versioning.mdx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+### How the SDK works with steps
+
+As covered in [_How Inngest functions are executed_](/docs/learn/how-functions-are-executed), each step in a function has a unique identifier, represented as a string. Each time a step is found, the SDK checks whether the step has been executed. It does this by:
+
+1. Hashing the step's unique identifier along with an counter of the number of times the step has been called. This allows steps to be used in a loop.
```
Comment (sylwiavargas) [inline pages/docs/learn/versioning.mdx:15]:
```suggestion
1. Hashing the step's unique identifier along with a counter of the number of times the step has been called. In this way, steps can be used in a loop.
```

File: pages/docs/learn/versioning.mdx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+## Change management across versions
+
+Given the above, there are a few strategies for change management:
+
+- **Adding new steps to a function is generally safe.** New steps will be executed when the functions re-run (after a step completes). If a function has step `[A, B, C]` and you add a new step `Z` in-between the first steps, the executor will run steps `[A, Z, B, C]` and log a warning. The caveat here is that you must take care to ensure that the new step can run out-of-order and doesn't reference undefined variables. Note that step `B` and `C` will *not* automatically re-run. Instead, a warning will be logged by default. You can change logging a warning and instead permanently fail by enabling strict mode. Permanently failing runs is acceptable, and you can use replay to bulk-replay permanent failures.
```
Comment (sylwiavargas) [inline pages/docs/learn/versioning.mdx:33]:
```suggestion
- **Adding new steps to a function is generally safe.** New steps will be executed when the functions re-run (after a step completes). Imagine a function has steps `[A, B, C]`. When you add a new step `Z` in-between the first steps, the executor will run steps `[A, Z, B, C]` and log a warning. The caveat here is that you must take care to ensure that the new step can run out-of-order and doesn't reference undefined variables. Note that step `B` and `C` will *not* automatically re-run. Instead, a warning will be logged by default. You can change logging a warning and instead permanently fail by enabling strict mode. Failing runs permanently is acceptable, and you can use [Replay](/docs/platform/replay) to bulk-replay permanent failures.
```

File: pages/docs/learn/versioning.mdx (hunk 1 above)
Comment (djfarrelly) [reply pages/docs/learn/versioning.mdx:5]:
```suggestion
ong-running functions inevitably change over time. Inngest enables developers to implement multiple strategies for changing long-running code over time. To manage these changes effectively, it's crucial to understand how the SDK implements determinism and [executes steps](/docs/learn/how-functions-are-executed).
```

File: pages/docs/learn/versioning.mdx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+- **Forcing steps to re-run by changing step IDs.** This changes the hash, which forces re-evaluation as the step's state is not found. Note that the SDK will log a warning by default as the order of step execution changes. If you change step `C`'s ID to `Z`, your run's state will expect steps `[A, B, C]` to run and instead will see `[A, B, Z]`.
```
Comment (sylwiavargas) [inline pages/docs/learn/versioning.mdx:34]:
```suggestion
- **Forcing steps to re-run by changing step IDs.** This changes the hash, which forces re-evaluation as the step's state is not found. Note that the SDK will log a warning by default as the order of step execution changes. If you change step `C`'s ID to `F`, your run's state will expect steps `[A, B, C]` to run and instead will see `[A, B, F]`.
```

changing so folks know this is not a continuation of the previous point

File: pages/docs/learn/versioning.mdx (hunk 3 above)
Comment (djfarrelly) [reply pages/docs/learn/versioning.mdx:15]:
```suggestion
1. Hashing the step's unique identifier along with a counter of the number of times the step has been called. This enables steps to be used in a loop.
```
Riff on the edit for clarity of what this does

**Review by sylwiavargas — APPROVED** [review]
Thank you!
Left a few comments, approving already!

**Pull request merged**